In [1]:
from dotenv import load_dotenv
import os
import sys

sys.path.append('../')

load_dotenv()

config = {
    "base_url": os.getenv("BASE_URL"),
    "client_id": os.getenv("CLIENT_ID"),
    "client_secret": os.getenv("CLIENT_SECRET")
}


In [2]:
import re

def get_page(url):
    """
    Devuelve el número de página de una URL.
    """
    match = re.search(r'page=(\d+)', url)
    if match:
        return int(match.group(1))


In [3]:
from services.fiscal_mx_service import FiscalMXService

fiscal_mx_service = FiscalMXService(config)
all_invoices = []
link_id = 'c4432b21-b248-4bff-ab46-05ec06a22da1'

page= 1

while page:
    invoices = fiscal_mx_service.get_invoices(link_id,page)
    all_invoices.extend(invoices['results'])
    page = get_page(invoices['next']) if invoices['next'] else None

In [4]:
import pandas as pd

df = pd.DataFrame(all_invoices)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 160 entries, 0 to 159
Data columns (total 37 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   id                          160 non-null    object 
 1   link                        160 non-null    object 
 2   created_at                  160 non-null    object 
 3   type                        160 non-null    object 
 4   folio                       0 non-null      object 
 5   usage                       160 non-null    object 
 6   status                      160 non-null    object 
 7   payroll                     160 non-null    object 
 8   version                     160 non-null    object 
 9   currency                    0 non-null      object 
 10  payments                    160 non-null    object 
 11  warnings                    0 non-null      object 
 12  sender_id                   160 non-null    object 
 13  tax_amount                  160 non

In [6]:
df = df.dropna(axis=1, how='any').copy()

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 160 entries, 0 to 159
Data columns (total 28 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       160 non-null    object 
 1   link                     160 non-null    object 
 2   created_at               160 non-null    object 
 3   type                     160 non-null    object 
 4   usage                    160 non-null    object 
 5   status                   160 non-null    object 
 6   payroll                  160 non-null    object 
 7   version                  160 non-null    object 
 8   payments                 160 non-null    object 
 9   sender_id                160 non-null    object 
 10  tax_amount               160 non-null    float64
 11  receiver_id              160 non-null    object 
 12  sender_name              160 non-null    object 
 13  collected_at             160 non-null    object 
 14  invoice_date             1

In [8]:
df.head()

id                                  link  \
0  67b1839f-13a9-4c07-a23d-cadf2645f05b  c4432b21-b248-4bff-ab46-05ec06a22da1   
1  c57021f2-4055-4d72-84b5-6782a8707b33  c4432b21-b248-4bff-ab46-05ec06a22da1   
2  1c0b33b0-9b06-4ccc-a354-44aca5b772b3  c4432b21-b248-4bff-ab46-05ec06a22da1   
3  72248a88-5ae1-4860-9e1c-d003f14264f1  c4432b21-b248-4bff-ab46-05ec06a22da1   
4  5d65feea-e140-496c-927d-085975a492a5  c4432b21-b248-4bff-ab46-05ec06a22da1   

                    created_at     type usage  status  \
0  2025-07-31T12:00:44.246920Z   INFLOW   G03  Vigent   
1  2025-07-31T12:00:44.246899Z   INFLOW   G03  Vigent   
2  2025-07-31T12:00:44.246878Z  OUTFLOW   G03  Vigent   
3  2025-07-31T12:00:44.246857Z  OUTFLOW   G03  Vigent   
4  2025-07-31T12:00:44.246835Z  OUTFLOW   G03  Vigent   

                                             payroll version  \
0  {'days': None, 'type': None, 'amount': 40.2, '...     3.3   
1  {'days': None, 'type': None, 'amount': 40.2, '...     3.3   
2  {'days': None, 'type': None, 'amount': 40.2, '...     3.3   
3  {'days': None, 'type': None, 'amount': 40.2, '...     3.3   
4  {'days': None, 'type': None, 'amount': 40.2, '...     3.3   

                                            payments      sender_id  ...  \
0  [{'date': '2025-06-21T09:38:27.254442', 'amoun...  PEVS00050734I  ...   
1                                                 []  REBF170804BSQ  ...   
2                                                 []  WONZ231224NY5  ...   
3  [{'date': '2025-05-08T18:24:52.975934', 'amoun...  QIMV571204G07  ...   
4                                                 []  HEAA771128VCX  ...   

   exchange_rate  receiver_name payment_method place_of_issue discount_amount  \
0              1  Receptor F967            PUE          11000             0.0   
1              1  Receptor 9264            PPD          11000          1278.0   
2              1  Receptor EFFA            PPD          11000         13044.0   
3              1  Receptor 8342            PPD          11000             0.0   
4              1  Receptor 1BAE            PPD          11000          9426.0   

                                     invoice_details subtotal_amount  \
0  [{'quantity': 1, 'unit_code': 'ACT', 'tax_amou...            0.00   
1  [{'quantity': 5, 'unit_code': 'E48', 'tax_amou...        62068.55   
2  [{'quantity': 9, 'unit_code': 'E48', 'tax_amou...        88995.27   
3  [{'quantity': 1, 'unit_code': 'ACT', 'tax_amou...            0.00   
4  [{'quantity': 1, 'unit_code': 'E48', 'tax_amou...        65730.90   

           certification_date               invoice_identification  \
0  2025-05-16T09:06:22.620431  4b11e71a-948a-3b5f-3634-9f783d3fc39   
1  2025-07-06T03:37:11.877383  fb12ad63-6c12-47bf-5565-b1e6b335a25   
2  2025-05-25T04:32:15.287479  fddc6c3d-ba7e-1627-72f7-fc8ae9314e5   
3  2025-06-28T02:01:29.781181  fe81c6c3-aa6c-ebc1-61db-38ad5d9fbce   
4  2025-07-13T08:50:14.313874  be296788-b2f2-1688-96a4-5497188d74e   

  certification_authority  
0           f224b774924b3  
1           dacdbadfabdfe  
2           2788274f4e78c  
3           9b3ee1e62c814  
4           fe5865f14f5fc  

[5 rows x 28 columns]

In [9]:
df.sender_name.value_counts()

sender_name
Emisor 2CD1    1
Emisor A642    1
Emisor E7A3    1
Emisor CEF3    1
Emisor F5B6    1
              ..
Emisor AF2E    1
Emisor 5921    1
Emisor 771B    1
Emisor 4413    1
Emisor 4D4A    1
Name: count, Length: 160, dtype: int64

In [11]:
df.receiver_name.value_counts()

receiver_name
Receptor F967    1
Receptor 9264    1
Receptor A8F6    1
Receptor B58C    1
Receptor 2955    1
                ..
Receptor E9B3    1
Receptor C242    1
Receptor 77B8    1
Receptor 518C    1
Receptor 34E8    1
Name: count, Length: 160, dtype: int64

In [12]:
df.type.value_counts()

type
INFLOW     86
OUTFLOW    74
Name: count, dtype: int64

In [13]:
df.invoice_type.value_counts()

invoice_type
Egreso      35
Traslado    34
Ingreso     32
Nómina      31
Pago        28
Name: count, dtype: int64

In [14]:
min(df.invoice_date), max(df.invoice_date)

('2022-01-01', '2025-08-30')

In [15]:
df.payment_method.value_counts()

payment_method
PUE    84
PPD    76
Name: count, dtype: int64

# Mexican Fiscal Invoices EDA Summary

## Dataset Overview
- **Total Records**: 160 invoices
- **Date Range**: January 1, 2022 to August 30, 2025
- **Data Quality**: After cleaning, 28 columns with complete data (removed 9 columns with null values)

## Invoice Distribution

### Transaction Flow
- **INFLOW**: 86 invoices (53.8%)
- **OUTFLOW**: 74 invoices (46.2%)

### Invoice Types
- **Egreso**: 35 invoices (22%)
- **Traslado**: 34 invoices (21%)
- **Ingreso**: 32 invoices (20%)
- **Nómina**: 31 invoices (19%)
- **Pago**: 28 invoices (18%)

### Payment Methods
- **PUE** (Pago en una sola exhibición): 84 invoices (52.5%)
- **PPD** (Pago en parcialidades o diferido): 76 invoices (47.5%)

## Key Characteristics
- **Unique Senders**: 160 different senders (each appears once)
- **Unique Receivers**: 160 different receivers (each appears once)
- **Data Completeness**: High quality dataset with consistent structure
- **Time Span**: Covers approximately 3.5 years of fiscal activity

## Business Insights
- Balanced mix of incoming and outgoing invoices
- Diverse invoice types covering various business operations (expenses, transfers, income, payroll, payments)
- Even split between immediate and deferred payment methods
- Each transaction involves unique sender-receiver pairs, suggesting diverse business relationships
- Strong data quality with complete information for all core fields after cleaning

## Technical Notes
- Original dataset: 37 columns, reduced to 28 after removing null columns
- All monetary amounts, dates, and identification fields are properly populated
- Data structure is consistent and ready for further analysis or modeling